pyi.py

```diff
@@ -917,8 +917,47 @@ def _check_union_members(self, members: Sequence[ast.expr]) -> None:
                 dupes_in_union = True
 
         if not dupes_in_union:
+            self._check_for_redundant_numeric_unions(members)
             self._check_for_multiple_literals(members)
 
+    def _Y041_error(
+        self, members: Sequence[ast.expr], subtype: str, supertype: str
+    ) -> None:
+        self.error(
+            members[0],
+            Y041.format(implicit_subtype=subtype, implicit_supertype=supertype),
+        )
+
+    def _check_for_redundant_numeric_unions(self, members: Sequence[ast.expr]) -> None:
+        complex_in_union, float_in_union, int_in_union = False, False, False
+
+        for member in members:
+            if isinstance(member, ast.Name):
+                name = member.id
+            elif (
+                isinstance(member, ast.Attribute)
+                and isinstance(member.value, ast.Name)
+                and member.value.id == "builtins"
+            ):
+                name = member.attr
+            else:
+                continue
```

Comment:
This is now pretty similar logic to lines 750-759, but I'm not sure it can be nicely generalised.

Reply:
Like this? Both instances would benefit from extraction, in my opinion, even as separate functions.

```python
def _extract_name(item: ast.AST, modules: Container[str]) -> str | None:
        if isinstance(item, ast.Name):
            return item.id
        elif (
            isinstance(item, ast.Attribute)
            and isinstance(item.value, ast.Name)
            and item.value.id in modules
        ):
            return item.attr
        else:
            return None
```

Reply:
Yeah, it was the naming that was causing me to hold off on this.

Reply:
I've had a go at factoring it out.